Common/cpp/SharedItems/Shareables.h

```diff
@@ -0,0 +1,270 @@
+      const std::shared_ptr<Scheduler> &_scheduler)
+      : rnRuntime(_rnRuntime), uiRuntime(_uiRuntime), scheduler(_scheduler) {}
+
+  std::shared_ptr<CoreFunction> valueUnpacker;
+
+  inline bool isUIRuntime(const jsi::Runtime &rt) const {
+    return &rt == uiRuntime;
+  }
+
+  inline bool isRNRuntime(const jsi::Runtime &rt) const {
+    return &rt == rnRuntime;
+  }
+
+  void scheduleOnUI(std::function<void()> job) {
+    scheduler->scheduleOnUI(job);
+  }
+
+  void scheduleOnJS(std::function<void()> job) {
+    scheduler->scheduleOnJS(job);
+  }
+};
+
+// Core functions are not allowed to capture outside variables, otherwise they'd
+// try to access _closure variable which is something we want to avoid for
+// simplicity reasons.
+class CoreFunction {
+ private:
+  std::shared_ptr<jsi::Function> rnFunction;
+  std::shared_ptr<jsi::Function> uiFunction;
+  std::string functionBody;
+  std::string location;
+  JSRuntimeHelper
+      *runtimeHelper; // runtime helper holds core function references, so we
+                      // use normal pointer here to avoid ref cycles.
+ public:
+  CoreFunction(
+      JSRuntimeHelper *_runtimeHelper,
+      const jsi::Value &workletObject);
+  jsi::Value call(jsi::Runtime &rt, const jsi::Value &arg0);
+};
+
+class Shareable {
+ protected:
+  virtual jsi::Value toJSValue(jsi::Runtime &rt) = 0;
+
+ public:
+  virtual ~Shareable();
+
+  enum ValueType {
+    UndefinedType,
+    NullType,
+    BooleanType,
+    NumberType,
+    // SymbolType, TODO
+    // BigIntType, TODO
+    StringType,
+    ObjectType,
+    ArrayType,
+    WorkletType,
+    RemoteFunctionType,
+    HandleType,
+  };
+
+  Shareable(ValueType valueType_) : valueType(valueType_){};
+  virtual jsi::Value getJSValue(jsi::Runtime &rt) {
+    return toJSValue(rt);
+  }
+
+  ValueType valueType;
+};
+
+class RetainingShareable : public Shareable {
+ public:
+  jsi::Runtime *hostRuntime;
+  std::shared_ptr<jsi::WeakObject> remoteValue;
+
+  RetainingShareable(jsi::Runtime &rt, ValueType valueType)
+      : Shareable(valueType), hostRuntime(&rt) {}
+
+  jsi::Value getJSValue(jsi::Runtime &rt) override final;
+};
+
+class ShareableJSRef : public jsi::HostObject {
+ public:
+  std::shared_ptr<Shareable> value;
+  ShareableJSRef(std::shared_ptr<Shareable> _value) : value(_value){};
+  virtual ~ShareableJSRef(){};
+
+  static jsi::Object newHostObject(
+      jsi::Runtime &rt,
+      const std::shared_ptr<Shareable> &value) {
+    return jsi::Object::createFromHostObject(
+        rt, std::make_shared<ShareableJSRef>(value));
+  }
+};
+
+std::shared_ptr<Shareable> extractShareableOrThrow(
+    jsi::Runtime &rt,
+    const jsi::Value &maybeShareableValue);
+
+class ShareableArray : public RetainingShareable {
+ public:
+  ShareableArray(jsi::Runtime &rt, const jsi::Array &array);
+
+  jsi::Value toJSValue(jsi::Runtime &rt) override {
+    auto ary = jsi::Array(rt, data.size());
+    for (size_t i = 0; i < data.size(); i++) {
```

Comment:
Not sure if `jsi::Array::size` is marked as `const` so this can be optimized away by the compiler.
```suggestion
    size_t size = data.size();
    auto ary = jsi::Array(rt, size);
    for (size_t i = 0; i < size; i++) {
```